sstables/sstable_set.cc

```diff
@@ -1163,6 +1175,75 @@ class time_series_reader_merger {
     }
 };
 
+flat_mutation_reader
+time_series_sstable_set::create_single_key_sstable_reader(
```

Comment:
I think I found an argument against putting this in sstable_set: memtables. If queries touch the recent end of the time series (common in CDC, and also common in many time series workloads) we might not need to open any sstable at all. So we should merge the streams at a higher level, when memtables are also considered a stream.

Rough proposal:
 - add min/max clustering key range accessors to flat_mutation_reader, implement for sstables/memtables
 - create a lazy_reader that wraps another reader but delays opening it until needed
 - the new combining reader accepts these wrapped readers (not knowing they are wrapped), pushes them into a heap by clustering order, consumes them as needed

Reply:
Maybe we can add these accessors to `mutation_source`, then we could avoid yet another wrapper. While at it we could fold [mutation_source_metadata](https://github.com/scylladb/scylla/blob/master/mutation_source_metadata.hh) into `mutation_source` too.

Reply:
Actually I remember wanting to fold `mutation_source_metadata` into `mutation_source` and ended up not doing it for some reason I can't remember. Which doesn't mean it cannot be done. Anyway this is unrelated.

Reply:
That makes sense - an sstable _is_ a mutation source.

It will cause an allocation to convert an sstable to mutation_source, but perhaps we can fix that by making mutation_source an abstract base class.

Reply:
Well we already have an abstract base-calss, we just pushed it out-of-sight with `std::function<>`. :)

Reply:
@avikivity 

The problem is that the min/max clustering key range is not exact, it only gives a lower bound and upper bound. The bounds may be achieved by a different partition than the one we're querying, or they might be set to -infinity/+infinity because the sstable was missing metadata.

To push a lazy_reader into a heap by clustering order, we need to know its first clustering key. To know the first clustering key, we **must** fetch the first fragment (because the min bound is not exact). For sstable readers the only way is to open the sstable.

Reply:
If queries touch the recent end of the time series, no sstables will be opened even before this change, they will be filtered out. `create_single_key_sstable_reader` will return a reader with an empty partition (returning only a `start_partition` and an `end_partition` fragment).

Reply:
> @avikivity
> 
> The problem is that the min/max clustering key range is not exact, it only gives a lower bound and upper bound. The bounds may be achieved by a different partition than the one we're querying, or they might be set to -infinity/+infinity because the sstable was missing metadata.

Is this a problem in practice? If the clustering key is time, it will be close to the timestamp, and so all clustering keys for sstables in the same window will have similar interval bounds.

> 
> To push a lazy_reader into a heap by clustering order, we need to know its first clustering key. To know the first clustering key, we **must** fetch the first fragment (because the min bound is not exact). For sstable readers the only way is to open the sstable.

The min bound isn't exact, but it will filter out all irrelevant sstables in most cases, and almost all irrelevant sstables in all cases.

Reply:
> If queries touch the recent end of the time series, no sstables will be opened even before this change, they will be filtered out. `create_single_key_sstable_reader` will return a reader with an empty partition (returning only a `start_partition` and an `end_partition` fragment).

You mean using `filter_sstable_for_reader_by_ck`? Yes.

Reply:
> Is this a problem in practice? If the clustering key is time, it will be close to the timestamp, and so all clustering keys for sstables in the same window will have similar interval bounds.

They probably will have similar bounds, so the min bound will be close to the first clustering key. But we need it to be **exact** for the lazy_reader proposal to give correct results.

Otherwise, the following can happen:
1. we put two readers into the heap, R1 and R2. R1's min lower bound says key `0`, R2's min lower bound says key `1`. So the heap sorts R1 before R2.
2. We take the min element from the heap, which is R1. We fetch the first fragment. It gives key `2` (the min bound was inexact, because the key `0` was in a different partition).
3. We take the min element from the heap, which is R2. We fetch the first fragment. It gives key `1`.
4. No profit, we returned keys in the wrong order.

> Why would an sstable lack metadata?

E.g. cause it comes from Cassandra or an old version. But it probably won't; it's just an example of what can happen so the bounds are inexact. The example with multiple partitions with different ranges is probably better.

Reply:
I guess we could do the following:

open the readers and fetch their first fragment until the first fragment has a smaller position than the min key of the next reader present in the heap.

But it complicates the solution, because we need to put the opened readers into another heap. So we got even more heaps.

That first heap (which uses min bounds) is exactly what `min_position_reader_queue` does (except that it uses a multimap). So we basically reinvented the PR we're discussing.

Reply:
So in the end, the lazy_reader solution would need the merger to know that the underlying readers are wrapped, because it needs to know that they can return the first fragment different (with a greater position) than what they initially claimed.

Reply:
> > Is this a problem in practice? If the clustering key is time, it will be close to the timestamp, and so all clustering keys for sstables in the same window will have similar interval bounds.
> 
> They probably will have similar bounds, so the min bound will be close to the first clustering key. But we need it to be **exact** for the lazy_reader proposal to give correct results.
> 
> Otherwise, the following can happen:
> 
> 1. we put two readers into the heap, R1 and R2. R1's min lower bound says key `0`, R2's min lower bound says key `1`. So the heap sorts R1 before R2.
> 2. We take the min element from the heap, which is R1. We fetch the first fragment. It gives key `2` (the min bound was inexact, because the key `0` was in a different partition).
> 3. We take the min element from the heap, which is R2. We fetch the first fragment. It gives key `1`.
> 4. No profit, we returned keys in the wrong order.
> 

Yes, you're right. Maybe we should have an option for TWCS to use the clustering key for segregation (and then it's exact).

Reply:
> I guess we could do the following:
> 
> open the readers and fetch their first fragment until the first fragment has a smaller position than the min key of the next reader present in the heap.

Opening the readers is already a loss. If you select from a 10-window range you now open 10 sstables instead of 1. That's a factor of 10.

There's a chance that the reader will be abandoned before the end, and even if not, it bloats the read memory consumption by a factor of 10 for all the readers that are opened but unused.

> 
> But it complicates the solution, because we need to put the opened readers into another heap. So we got even more heaps.
> 
> That first heap (which uses min bounds) is exactly what `min_position_reader_queue` does (except that it uses a multimap). So we basically reinvented the PR we're discussing.

Yes, I'm not interested in optimizing the compares, I'm interested in having fewer readers open at the same time.

Reply:
>  Maybe we should have an option for TWCS to use the clustering key for segregation (and then it's exact).

It's something to think about. But we probably still want to optimize without this option though (do we?) so I'd like to push this PR forward.

Also note that the clustering key segregation would, in general, require to put different partitions into different sstables when flushing a single memtable. Something like the object compaction strategy thingy from the hackaton.

Reply:
> > Maybe we should have an option for TWCS to use the clustering key for segregation (and then it's exact).
> 
> It's something to think about. But we probably still want to optimize without this option though (do we?) so I'd like to push this PR forward.

But, aren't we blocked? We have to open all sstables up front. Once we do that, we may as well let combining_reader to the combining.

> 
> Also note that the clustering key segregation would, in general, require to put different partitions into different sstables when flushing a single memtable. Something like the object compaction strategy thingy from the hackaton.

TWCS already does that, fragements that belong to different windows are flushed into separate sstables.

Reply:
> But, aren't we blocked? We have to open all sstables up front.

No, why? The entire purpose of this PR is to not open **all** sstables up front.
We may need to open some of them earlier than they will actually return a fragment (because the bounds are inexact), but it's not opening all of them.

Consider the following example. We have the following readers with the following bounds and keys for the queried partition:
- R1 has bounds [0, 10], but keys [3, 7] for the given query
- R2 has bounds [1, 13], and keys [5, 9] for the query
- R3 has bounds [15, 20] and keys [16, 17] for the query

The reader will do the following:
1. open R1
2. open R2
3. read and merge from R1 and R2
4. after key 7, close R1, but continue reading from R2 with no merging (it will gallop)
6. open R3, read, close

so it does have R1 and R2 opened while only R1 is returning fragments (because R2 really starts at key 5, but the bound was inexact), but it does improve things a lot.
In particular, if the bounds are close to the exact bounds for the given query (and they probably often will), it will give close-to-optimal results.

> TWCS already does that, fragements that belong to different windows are flushed into separate sstables.

but in the current window you have multiple sstables with overlapping clustering ranges and overlapping partition key ranges (it uses STCS). The segregation you mentioned would have to be performed on memtable flush, as the S3 project does

Reply:
But above you said we cannot rely on the bounds because they might result in the wrong sort order. What will happen in this case:

    R1 has bounds [0, 10], but keys [6, 7] for the given query
    R2 has bounds [1, 13], and keys [5, 9] for the query
    R3 has bounds [15, 20] and keys [16, 17] for the query

Here, the bound based order (R1, R2, R3) is incorrect, as the keys would result in a different sort order (R2, R1, R3).

Reply:
Here:
```
        auto next_peeked_pos = _peeked_readers.empty() ? _pr_end : _peeked_readers.front().fragment.position();
        // There might be queued readers with min_position <= next_peeked_pos:
        if (!_reader_queue.empty(next_peeked_pos)) {
```
I keep opening readers until I see there are no more readers that could potentially sort before the readers that I have already opened.

So the queue would first return R1, I would fetch the first fragment from R1, and this code would get `next_peeked_pos = 6`. It would query the queue `_reader_queue.empty(6)`, which would return `false`, so I would `pop` from the queue, which will return R2. I will peek R2, it will return the `5` fragment, which will get inserted into the heap. I will enter this code again, this time `next_peeked_pos = 5`, and `reader_queue.empty(5)` will return `true`, so I will enter the `pop_heap` loop.

Reply:
I do rely on the bounds in that I rely that hey are **bounds**. I.e. I rely on the fact that `min_position() <= real_min_position()`, where `real_min_position()` is the position of the first fragment that the reader will return. I don't rely on this for correctness, only for optimization.

Reply:
Then I don't see a problem in integrating this into `mutation_source`, instead of a new abstraction, the `lazy_reader`.

Reply:
I have to admit I don't really get the idea with `mutation_source` and what problem it solves.

The problem that this PR solves is to open fewer sstables in parallel when querying a clustering range in a single partition,  if it's possible to say that some sstables do not have to be opened yet. The assumption we make is that the clustering keys are increasing over time, so the optimization makes sense (it really results in fewer sstables being opened in parallel). In effect we reduce memory consumption (and potentially reduce the number of comparisons during heap operations since there are fewer fragments sitting in the heap at a time, but that's probably not the most important thing here).

What exactly is this new generalized proposal, and what is the goal? I understand that we somehow want to incorporate memtables into the sstable-merging algorithm, is that right? The motivation Avi gave was that "we might not need to open any sstable at all" if all the sstables get filtered when querying the tail of the clustered range, but that will already happen with this PR (and it even happens today without it). Hence I don't see how it motivates this generalization, the generalization won't change anything in that matter.

Reply:
I have to admit I lost track of this discussion, I just realized that it was @avikivity who proposed the lazy reader.

Avi's idea was to use the already existing `mutation_source` abstraction to implement incremental selection. The `mutation_source` would get two new methods: `min_clustering_key` and `max_clustering_key` and the new merger would simply use these instead of using the sstable set directly.

Reply:
Let's zoom out. I'm sure I'm repeating things we all know but let's establish something we can agree on.

Two queries:

1.    `SELECT * FROM tab WHERE pk = ? AND ck > :lower AND CK < :upper`

(or similar half-bounded queries). For this case, with the additional constraint that the ck range is small compared to a window, we already do well. We will filter out unneeded sstables, and the sstables we do open need to be opened in parallel since they contain the same ck intervals.

2.    `SELECT * FROM tab WHERE pk = ?`

(or half-bounded or fully bounded ck intervals, where the ck interval is larger than a window) 

In this case, the number of sstables needed for a page is much smaller than the number of sstables needed for the entire query. So opening (and closing) them incrementally is a benefit, potentially a large benefit if the interval is very large.

 Let's say for sstable X, the minimum ck is X_min_ck. Then we can defer opening it until we've seen in our stream a ck > X_min_ck. Once we do, we need to open X and merge it into the stream, possibly before ck.

Reply:
> For this case, with the additional constraint that the ck range is small compared to a window, we already do well. We will filter out unneeded sstables, and the sstables we do open need to be opened in parallel since they contain the same ck intervals.

If the ck range is in the current window and there were a bunch of sstable flushes in the current window then it's possible that we open sstables in parallel unnecessarily (the different flushes produce sstables that are mostly disjoint). This PR helps with that problem.

> In this case, the number of sstables needed for a page is much smaller than the number of sstables needed for the entire query. So opening (and closing) them incrementally is a benefit, potentially a large benefit if the interval is very large.

Yes.

> Let's say for sstable X, the minimum ck is X_min_ck. Then we can defer opening it until we've seen in our stream a ck > X_min_ck. Once we do, we need to open X and merge it into the stream, possibly before ck.

Yes, with the small inconvenience that we don't know `X_min_ck`. But we do know something: we know `X_ck_lower_bound`, where `X_ck_lower_bound <= X_min_ck`, and with high probability the range `(X_ck_lower_bound, X_min_ck)` is small.
We can defer opening `X` until we've seen in our stream a ck `> X_ck_lower_bound`. This is what the PR does.

Reply:
> > For this case, with the additional constraint that the ck range is small compared to a window, we already do well. We will filter out unneeded sstables, and the sstables we do open need to be opened in parallel since they contain the same ck intervals.
> 
> If the ck range is in the current window and there were a bunch of sstable flushes in the current window then it's possible that we open sstables in parallel unnecessarily (the different flushes produce sstables that are mostly disjoint). This PR helps with that problem.

btw, pedantic terminology note: the sstables are always open, we open more sstable readers.

In what way does it help that normal clustering key range filtering doesn't?

> 
> > In this case, the number of sstables needed for a page is much smaller than the number of sstables needed for the entire query. So opening (and closing) them incrementally is a benefit, potentially a large benefit if the interval is very large.
> 
> Yes.
> 
> > Let's say for sstable X, the minimum ck is X_min_ck. Then we can defer opening it until we've seen in our stream a ck > X_min_ck. Once we do, we need to open X and merge it into the stream, possibly before ck.
> 
> Yes, with the small inconvenience that we don't know `X_min_ck`. But we do know something: we know `X_ck_lower_bound`, where `X_ck_lower_bound <= X_min_ck`, and with high probability the range `(X_ck_lower_bound, X_min_ck)` is small.
> We can defer opening `X` until we've seen in our stream a ck `> X_ck_lower_bound`. This is what the PR does.

Well, then I forget what the controversy was (I don't even remember controversy, I just assume it from the amount of time it takes to load this page). Maybe we're all agreeing here.

Reply:
> In what way does it help that normal clustering key range filtering doesn't?

It helps when the ck range in the current window covers multiple sstables (which may happen if there were multiple flushes during the current time window).
Before PR all these sstables (their readers) would be kept open during the range query, after PR they will be opened when needed and closed as soon as possible.
Ck filtering only helps by filtering out previous time windows. The PR helps with sstables that were not filtered out.

> Well, then I forget what the controversy was (I don't even remember controversy, I just assume it from the amount of time it takes to load this page). Maybe we're all agreeing here.

The question is whether I should create more abstractions and interfaces that generalize the merging algorithm to `mutation_source`s or whatever, or if the architecture proposed in this PR is good enough (the logic is encapsulated and private to `time_series_sstable_set::create_single_key_sstable_reader`). If there's a push towards generalization and abstraction, my question is why (what problem does it solve).

Reply:
I think the current architecture is fine. It is not how I would have done it, I tried to convince you to agree with me and failed, but that is not grounds for rejecting a patch-set.

Reply:
Thanks. I'll add for consolation: look how easy it was to generalize `combined_mutation_reader` so it could fit this new use case. It was just adding a template parameter, the change took like what, 20 LOC in total (not counting the code movement)?

If in the future we find a need to generalize `min_position_reader_queue`+`time_series_reader_merger` for other use cases, it will be the same kind of refactor. It's just creating an interface over `min_position_reader_queue` and using it inside `time_series_reader_merger` + cutting and pasting code.